kubevirt-vmware/pkg/controller/ovirtprovider/ovirtprovider_controller.go

```diff
@@ -0,0 +1,288 @@
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileOVirtProvider) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling OVirtProvider")
+
+	// Fetch the OVirtProvider instance
+	instance := &kubevirtv1alpha1.OVirtProvider{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			reqLogger.Info("The request object cannot be found.")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		reqLogger.Info("Error reading the request object, requeuing.")
+		return reconcile.Result{}, err
+	}
+
+	connectionSecret, err := r.fetchSecret(instance)
+	if err != nil {
+		reqLogger.Error(err, "Failed to get Secret object for the oVirt connection")
+		return reconcile.Result{}, err // request will be re-queued
+	}
+	reqLogger.Info("Connection secret retrieved.")
+
+	if len(instance.Spec.Vms) == 0 { // list of oVirt VMs is requested to be retrieved
+		err = readVmsList(r, request, connectionSecret)
+		if err != nil {
+			reqLogger.Error(err, "Failed to read list of oVirt VMs.")
+			// TODO: check the reason to fail and decide whether to wait like for vmware or to fail the import
+		}
+
+		return reconcile.Result{}, err // request will be re-queued if failed
+	}
+
+	// secret is present, list of VMs is available, let's check for  details to be retrieved
+	var lastError error = nil
+	for _, vm := range instance.Spec.Vms { // sequential read is probably good enough, just a single VM or a few of them are expected to be retrieved this way
+		if vm.DetailRequest {
+			err = readVMDetail(r, request, connectionSecret, vm.Name)
+			if err != nil {
+				reqLogger.Error(err, fmt.Sprintf("Failed to read '%s' vm details.", vm.Name))
+				lastError = err
+			}
+		}
+	}
+
+	return reconcile.Result{}, lastError
+}
+
+func (r *ReconcileOVirtProvider) fetchSecret(provider *kubevirtv1alpha1.OVirtProvider) (*corev1.Secret, error) {
+	secret := &corev1.Secret{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Name: provider.Spec.Connection, Namespace: provider.Namespace}, secret)
+	return secret, err
+}
+
+func getClient(ctx context.Context, secret *corev1.Secret) (*Client, error) {
+	return NewClient(ctx, string(secret.Data["apiUrl"]), string(secret.Data["username"]), string(secret.Data["password"]), string(secret.Data["cluster"]))
+}
+
+// read whole list at once
+func readVmsList(r *ReconcileOVirtProvider, request reconcile.Request, connectionSecret *corev1.Secret) error {
```

Comment:
while do you pass `r *ReconcileOVirtProvider` as an argument instead of defining this method to be a function of  ReconcileOVirtProvider struct ?

Same goes for other methods below.

Reply:
Will update.